src/rest/actions/action.hpp

```diff
@@ -0,0 +1,280 @@
+    /**
+     * Get the status of the action.
+     *
+     * @return  The status of the action.
+     */
+    ActionStatus GetStatus(void) const { return mStatus; }
+
+    /**
+     * Get the status of the action as a string.
+     *
+     * @return  The status of the action as a string.
+     */
+    const char *GetStatusString(void) const { return StatusToString(mStatus); }
+
+    /**
+     * Is the action pending or active?
+     *
+     * @retval True  If the action is pending or active.
+     * @retval False If the action is not pending or active.
+     */
+    bool IsPendingOrActive(void) const { return mStatus == kActionStatusPending || mStatus == kActionStatusActive; }
+
+    /**
+     * Convert the status of the action to string.
+     *
+     * @param[in] aStatus  The status of the action.
+     * @return  The status of the action as a string.
+     *
+     */
+    static const char *StatusToString(ActionStatus aStatus);
+
+protected:
+    static constexpr Seconds kDefaultTimeout{60};
+
+    /**
+     * Attempts to read the timeout property in a actions attribute cJSON
+     * object.
+     *
+     * If the timeout property cannot be found aTimeout remains unchanged.
+     *
+     * @param       aJson     The attributes cJSON object.
+     * @param[out]  aTimeout  Written with the timeout value if found.
+     *
+     * @retval  OT_ERROR_NONE       The timeout property was found and written to aTimeout
+     * @retval  OT_ERROR_PARSE      The timeout property was found but could not be parsed
+     * @retval  OT_ERROR_NOT_FOUND  The timeout property was not found or could not be parsed
+     */
+    static otError ReadTimeout(const cJSON *aJson, Seconds &aTimeout);
+
+    /**
+     * Attempts to read the timeout property in a actions attribute cJSON
+     * object.
+     *
+     * If the timeout property cannot be found or parsed aDefault is returned.
+     *
+     * @param       aJson         The attributes cJSON object.
+     * @param[in]   aDefault      The default value to return if the timeout property is not found.
+     *
+     * @return  The timeout value if found, otherwise aDefault.
+     */
+    static Seconds ReadTimeoutOrDefault(const cJSON &aJson, Seconds aDefault);
+
+    /**
+     * Attempts to read the destination property in a actions attribute cJSON
+     *
+     * @param       aJson     The attributes cJSON object.
+     * @param[out]  aType     Written with the address type if found.
+     *
+     * @return  The destination property string if found, nullptr otherwise.
+     */
+    static const char *ReadDestination(const cJSON *aJson, AddressType &aType);
+
+    /**
+     * Adds the timeout entry to the attributes objects.
+     *
+     * @param aAttributes  A cJSON object to which the timeout property should be added
+     */
+    void JsonifyTimeout(cJSON *aAttributes) const;
+
+    /**
+     * Set the relationship to the result of this action.
+     */
+    void SetResult(const char *aType, std::string aUuid);
```

Comment:
If not mutable, perhaps use `const std::string &`?
```suggestion
    void SetResult(std::string &aType, std::string &aUuid);
```

Reply:
changed. thx.